test/with_dummyserver/test_socketlevel.py

```diff
@@ -843,6 +843,78 @@ def socket_handler(listener):
         pool = HTTPConnectionPool(self.host, self.port, retries=False)
         pool.request('GET', '/', headers=HTTPHeaderDict(headers))
         self.assertEqual(expected_headers, parsed_headers)
+    
+    def test_request_headers_are_sent_in_the_original_order(self):
+        # NOTE: Probability this test gives a false negative is 1/(K!)
+        K = 16
+        # NOTE: Provide headers in non-sorted order (i.e. reversed)
+        #       so that if the internal implementation tries to sort them,
+        #       a change will be detected.
+        expected_request_headers = [('X-Header-%d' % i, str(i)) for i in reversed(range(K))]
+        
+        headers = {'foo': 'bar', 'bAz': 'quux'}
+        actual_request_headers = []
+
+        def socket_handler(listener):
+            sock = listener.accept()[0]
+
+            buf = b''
+            while not buf.endswith(b'\r\n\r\n'):
+                buf += sock.recv(65536)
+
+            headers_list = [header for header in buf.split(b'\r\n')[1:] if header]
+
+            for header in headers_list:
+                (key, value) = header.split(b': ')
+                if not key.decode().startswith('X-Header-'):
```

Comment:
Naked `decode` calls make me very nervous because they can mysteriously break on systems with weird locales. Can we explicitly use `ascii` here?

While we're doing it, let's change the header check to use a unicode string (`u'X-Header-'`) to avoid Python 2 doing an implicit coercion and maybe getting this wrong again.